engine/src/main/java/org/terasology/registry/DynamicInstanceProvider.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.registry;
+
+import org.terasology.context.Context;
+import org.terasology.naming.Name;
+
+/**
+ * A DynamicInstanceProvider provides an instance dynamically.
+ */
+public interface DynamicInstanceProvider<T> {
+    /**
+     * Returns the instance for the given module.
+     *
+     * @param moduleId The module ID.
+     * @param context The current context.
+     * @return An instance.
+     */
+    T getInstanceForModule(Name moduleId, Context context);
```

Comment:
I think Context should not be an argument, but something that gets passed to the constructor when the object gets created. Reasoning: The provider will live for the same duration like the objects it provides.